Shorten getting started

mbolivar opened the pull request:
The main purpose of this PR is to make the getting started guide shorter and easier to follow for new users, especially now that we have gotten over the conversion to west.

Along the way, I add a generic page on managing environment variables and link to it from pages (including getting started ones) instead of reproducing bits and pieces of the relevant information.

mbolivar-nordic committed:
doc: unify documentation on setting environment variables

There are a few different places where alternatives for setting
environment variables are described. None of them is 100% complete, so
the results are likely to be confusing.

Make a single page on setting environment variables, how the zephyrrc
files work, how the zephyr-env scripts work, and some of the important
environment variables, with appropriate references elsewhere. (This is
inspired by the Arch wiki's excellent page on installing programs.)

Link to it from the getting started and application development pages
instead of repeating the information. This has the benefit of
shortening the getting started guide a bit more.

Add some concrete advice on checking the toolchain environment
variables in particular. This is a stumbling block for beginners.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -51,27 +50,21 @@ First, install the ``west`` binary and bootstrapper:
-.. note::
-   See :ref:`gs_python_deps` for additional clarfication on using the
-   ``--user`` switch.
+Some notes on the ``--user`` option:
```
My west autocomplete PR (#16013) will force you to rewrite this a bit

mbolivar replied:
I know; that's fine. I wanted to get this up before hopping on the plane and plan on rebasing once your PR is in.

carlescufi commented on doc/guides/env_vars.rst:
```diff
@@ -0,0 +1,144 @@
+
+   choco install rapidee
+
+You can then run ``rapidee`` from your terminal to launch the program.
```
RapidEE needs to be run as an administrator in order to save the changes you make to global env vars

mbolivar replied:
Maybe I can just recommend saving to user variables only here. OK for you?

mbolivar replied:
I am happy to add a `setx` example and think this would be an improvement to this document. Thank you for the reference.

carlescufi replied:
@mbolivar yes, please do document both RapidEE and `setx` and, in the case of RapidEE, let's save them as user variables. It's a bit confusing for the user since they show in a different panel from the system-wide ones, but the trade-off is worth it IMHO.
@marc-hb thanks for bringing up `setx`, I had forgotten this. And yes, RapidEE is great :)

marc-hb replied:
I didn't know about `setx` until a couple days ago! I would have use the GUI cause it's Windows :-)
More seriously, what drew my attention is giving the impression that Windows requires third-party software to change environment variables, that doesn't feel right.

marc-hb commented on doc/guides/env_vars.rst:
```diff
@@ -0,0 +1,144 @@
+   # Linux and macOS
+   export MY_VARIABLE=foo
+
+   # Windows
```
Just cmd.exe, not powershell

mbolivar replied:
I'm not going to fix this one though, because as stated above, it's a commonplace in our documentation that this means cmd.exe.

marc-hb commented on doc/getting_started/index.rst:
```diff
@@ -116,8 +95,7 @@ and more on how ``west`` helps manage multiple repositories.
 
-Next, install additional Python packages required by Zephyr in a shell or
-``cmd.exe`` prompt:
+Install Python packages required by Zephyr:
```
"Less is more" yet cleaning and deleting stuff is a difficult job that rarely gets the credit it deserves. Impressive dedication @mbolivar.

This being said, pip deserves a bit more introduction. So here's a suggested addition:

> [`pip`](https://pypi.org/project/pip) is Python's package installer. It first tries to re-use packages and package dependencies already installed by your operating system. If the requested versions are not found then it downloads them from the Internet and installs them in the .local/ subdirectory of your HOME. Note the package versions requested by Zephyr may conflict with other versions requested by other software, in that case you may want to look at [virtualenv](https://virtualenv.pypa.io/en/latest/)

pip is extremely important for Zephyr. It manages a large and crucial part of Zephyr dependencies and can puzzle users when it fails. This seems fortunately rare, except and as previous observed by @mbolivar,  Zephyr's `requirements.txt` files (there's another one in ci-tools, to be merged [one way](https://github.com/zephyrproject-rtos/zephyr/pull/15992) or the other) is very picky which could clash with other software. For security reasons this guide should not hide that pip downloads a ton of code from random places on the Internet. Right now there's not even a link to pip's documentation!
See some previous discussion at: https://lists.zephyrproject.org/g/users/message/1308 and elsewhere.

mbolivar replied:
I think you're right that this is important information that should be present in our documentation. I want to balance this with brevity in a 1,2,3 style getting started guide, however, where we are intentionally glossing over many details that are important in favor of getting people going.

How would you feel about a short user guide on pip, linked to from the getting started guide?

carlescufi replied:
I agree with @marc-hb, pip deserves at least a paragraph in our documentation. I think @mbolivar 's suggestion of a linked to short guide hits the sweet spot between brevity and thoroughness.

marc-hb replied:
While we seem to agree zero information and zero link about pip is not enough, I don't have enough pip information for a standalone page either. If my 4 sentences above are too much then take only the first half or maybe even less (an introduction and pointer to another, standalone page wouldn't be much shorter anyway)

mbolivar replied:
Given this feedback, I'm going to put Marc's content into a footnote, and a link to https://packaging.python.org/tutorials/installing-packages/ inline in the text.

marc-hb reviewed (commented):
requesting the addition (sorry!) of a very short pip introduction

mbolivar commented on doc/getting_started/toolchain_3rd_party_x_compilers.rst:
```diff
@@ -63,24 +70,29 @@ Intel ISSM
-      # Windows in cmd.exe
-      set ZEPHYR_TOOLCHAIN_VARIANT=issm
-      set ISSM_INSTALLATION_PATH=c:\issm0-toolchain-windows-2017-01-25
+      # Windows
```
We use `# Windows` ubiquitously to mean "windows cmd.exe", and are clear in the windows setup docs that we mean cmd.exe.

I'll leave this in to serve as a reminder of that, though.

mbolivar commented on doc/guides/env_vars.rst:
```diff
@@ -0,0 +1,144 @@
+The value will be lost if you close the window, etc.; run ``source
+zephyr-env.sh`` again to get it back.
+
+**Windows:**
```
See my other responses

mbolivar reviewed (commented):
Thanks for your thorough review, @marc-hb! I'll make some fixes soon and have some other responses.

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -80,29 +73,15 @@ Next, clone the Zephyr source code repositories from GitHub using the
-   working directory.
-
-.. note::
-   If you had previously cloned the zephyr repository manually using Git,
```
I disagree with removing `-l` here. I think it has a lot of value, not only for CI systems that come with their own Git client that will pull the repo CI is supposed to run on, but also for users that might want to clone manually.

mbolivar replied:
I think that setting up CI is an advanced use case not suitable for the getting started guide, as is cloning the multiple repositories we require manually. (Cloning by hand is also already served by https://docs.zephyrproject.org/latest/guides/west/without-west.html#no-west, which is linked to above.)

Documenting `west init -l` was all well and good while we were converting people to west, but are these use cases really significant enough to document here? The west multirepo page does document this option clearly in the "Main Commands" section.

carlescufi replied:
Fair enough, you've both convinced me. I am always reluctant to remove information, it takes real courage to do so.

carlescufi commented on doc/application/index.rst:
```diff
@@ -504,8 +503,9 @@ again.
 
 .. note:: The ``run`` target will use the QEMU binary available from the Zephyr
```
This is only true if building with `ZEPHYR_TOOLCHAIN_VARIANT=zephyr` AFAIK. On macOS for example, you can `ninja run` and it will take brew's QEMU.

mbolivar replied:
Isn't that implied by the unavailability of the SDK on macOS?

mbolivar replied:
Yes. If you set `ZEPHYR_SDK_INSTALL_DIR`, the build system tries to use the host tools from the SDK regardless of if the toolchain is obtained from the SDK. See the Linux page in the getting started guide.

marc-hb replied:
OK but is QEMU considered a "host tool" from that ZEPHYR_SDK_INSTALL_DIR perspective? Not according to @carlescufi initial comment above.

carlescufi replied:
@marc-hb it is, in the sense that if `ZEPHYR_SDK_INSTALL_DIR` is set, then the `qemu` binary will be executed from there, but if not it defaults to whatever `qemu` is in the `PATH`. This is the same for `dtc`.
@mbolivar I would still add a "if the Zephyr SDK is installed"

mbolivar replied:
> @mbolivar I would still add a "if the Zephyr SDK is installed"

I think this whole debate is a bit outside this PR's scope, but I'll add a patch on top to make this clearer while we are editing this paragraph.

marc-hb replied:
I think there are nuances still not captured by this note, however there's only so much you can put in a note like this and the addition of the hyperlink to the SDK page clarifies what "installed" means if desired which seems like the best solution. The only statement I would like added (on either page) is the confirmation that QEMU is indeed a "host tool" but no big deal.

carlescufi commented on doc/application/index.rst:
```diff
@@ -1100,7 +1101,8 @@ Make sure to follow these steps in order.
      will be checked for and used next.
 
-   - If an environment variable ``BOARD`` is set, its value will then be used.
+   - If an :rev:`environment variable <env_vars>` ``BOARD`` is set, its value
```
s/rev/ref

mbolivar replied:
Whoops

carlescufi commented on doc/guides/env_vars.rst:
```diff
@@ -0,0 +1,144 @@
+
+On macOS and Linux, add the ``export MY_VARIABLE=foo`` line to your shell's
+startup script in your home directory. For Bash, this is usually
+:file:`~/.bashrc`.  The setting will then be available whenever you open a new
```
On macOS it's typically `~/.bash_profile`, I would mention this here.

mbolivar replied:
Great, thanks!

marc-hb replied:
One of too many stackoverflow duplicates trying and failing to summarize this giantic mess: https://superuser.com/questions/183870/difference-between-bashrc-and-bash-profile

mbolivar replied:
I just noticed this comment -- what would you like me to do?

marc-hb replied:
I think changing just this sentence would be great:
```diff
- The setting will then be available whenever you open a new terminal window.
+ Changes in these startup scripts don't affect shell instances already started.
```

mbolivar replied:
Done (with a suggestion to open a new window added after a semicolon)

marc-hb replied:
I think this paragraph still gives the user too much confidence that everything will just work the same on any Linux distribution. There wouldn't be so many duplicates of https://superuser.com/questions/183870/difference-between-bashrc-and-bash-profile if it were the case, just check the number of links there or perform a quick search.
Merely adding some adverb ("generally" or "in most cases",...) to the second sentence would be enough to temper it.

mbolivar replied:
will change to: " try opening a new terminal window to get the new settings."

marc-hb replied:
I dove deeper into the /`etc/profile/.bashrc/bash_profile/`etc. abyss to understand why it was so difficult to unset Clear Linux' default CFLAGS (PR #17029) It's even worse than I expected. Not because of bash itself but because of what the Linux distributions make of it. Here's for instance how Fedora and Clear Linux source default files. Ubuntu is of course completely different again. Longer version at https://superuser.com/a/1452523/111302

```
ssh fedora29
 └─ -bash # login shell
      ├── /etc/profile
      |    ├─ /etc/profile.d/*.sh
      |    ├─ /etc/profile.d/sh.local
      |    └─ /etc/bashrc
      ├── ~/.bash_profile
      |    └─ ~/.bashrc
      |          └─ /etc/bashrc
      |
      |
      └─ $ bash  # non-login shell
            └─ ~/.bashrc
                 └─ /etc/bashrc
                       └─ /etc/profile.d/*.sh

ssh clearlinux
 └─ -bash # login shell
      ├── /usr/share/defaults/etc/profile
      |    ├─ /usr/share/defaults/etc/profile.d/*
      |    ├─ /etc/profile.d/*
      |    └─ /etc/profile
      ├── ~/.bash_profile
      |
      |
      └─  $ bash   # non-login shell
           ├─ /usr/share/defaults/etc/bash.bashrc
           |      ├─ /usr/share/defaults/etc/profile
           |      |    ├─ /usr/share/defaults/etc/profile.d/*
           |      |    ├─ /etc/profile.d/*
           |      |    └─ /etc/profile
           |      └─ /etc/profile
           └─ ~/.bashrc
```

carlescufi commented on doc/guides/env_vars.rst:
```diff
@@ -0,0 +1,144 @@
+
+.. code-block:: console
+
+   cat >>~/.zephyrrc <<EOF
```
I truly dislike here-documents, they are confusing for the user. Instead I would ask the user to create a file `~/.zephyrrc` with their favorite editor and then display the contents of it.

carlescufi reviewed (changes requested):
Thanks for this, especially the new section on env vars, it's a great step forward in clarity and helpfulness to new users. I've added some requests for changing bits here and there that I don't agree with.

mbolivar-nordic committed:
doc: periodic updates to getting started guide

Updates based on discussion and changes in supported features.

- Make the guide shorter by removing content that's not relevant to
  most users who are truly just getting started, such as information
  about pre-LTS versions that did not support west, and by being more
  concise in some places.

- Decrease the number of colored boxes. At the latest TSC F2F, the
  "note / warning / note / tip" contents were identified as a
  readability problem.

- Add additional information based on new west features, like "west
  boards".

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar commented:
I've posted a v2.

Thanks for the detailed review @carlescufi and @marc-hb; comments addressed.

mbolivar commented:
v3 has a trivial build fix and a bit more removed text.

marc-hb commented on doc/getting_started/toolchain_3rd_party_x_compilers.rst:
```diff
@@ -65,24 +70,29 @@ Intel ISSM
-   For example:
+#. To check that you have set these variables correctly in your current
+   environment, follow these example shell sessions (the
+   :envvar:`ISSM_INSTALLATION_PATH` values may be different on your system):
```
This introduction is too verbose, even more so now that there's an entire page about environment variables. It can be replaced with a single word: "Examples:"

mbolivar replied:
I think a plain "examples" would easily be misconstrued by the reader to mean "examples of how to set these variables" (which they are not ), rather than "examples of how to check you have set these variables correctly" (which they are).

Setting up a toolchain is the first time zephyr users are asked to set environment variables. I have watched users stumble over this process again and again, both in person and on IRC. It's probably one of the most common support issues for beginners, so as a special case, I am trying to make it extra clear how they can check their work.

Just FYI, this PR is also a stopgap and prep work for moving the getting started toolchain configuration to west. The content in this section will eventually be replaced with something like this:

```
west config build.zephyr_toolchain_variant issm
west config build.issm_installation_path /home/you/Downloads/issm0-toolchain-windows-2017-02-07
```

I am just trying to help avoid support requests in the meantime.

If you feel strongly that this is too verbose I'll just delete these "To check..." entirely, but I think that's a bad idea for reasons explained above.

dbkinder replied:
Perhaps it's time for a short usability test with some new Zephyr users to verify these new getting started documents are doing their job.

carlescufi replied:
I think that's a great idea, and one that perhaps @Thea-Aldrich could follow-up on?

Thea-Aldrich replied:
That's a great idea @carlescufi. I have a couple folks I can reach out to in the community who are interested in providing feedback. @dbkinder would you like to team up on drafting the questions and super informal methodology ya'll would like me to use as I gather feedback?

mbolivar replied:
> Perhaps it's time for a short usability test with some new Zephyr users to verify these new getting started documents are doing their job.

@dbkinder I agree, but wonder if that should block the merge or if we can get a more explicit verdict from you on the point being discussed here?

dbkinder replied:
Gosh no - don't block this PR merge for a usability test (we can update as needed later).

dbkinder replied:
BTW, there actually is a usability study going on. Let's get this PR merged and we'll address issues found during this study in a subsequent PR.

marc-hb commented on doc/getting_started/toolchain_3rd_party_x_compilers.rst:
```diff
@@ -109,18 +119,22 @@ You can build toolchains from source code using crosstool-NG.
-   For example:
+#. To check that you have set these variables correctly in your current
+   environment, follow these example shell sessions (the
+   :envvar:`XTOOLS_TOOLCHAIN_PATH` values may be different on your system):
```
too verbose, same as previous comment

marc-hb commented on doc/getting_started/toolchain_other_x_compilers.rst:
```diff
@@ -29,20 +29,24 @@ Follow these steps to use one of these toolchains.
-   Continuing the above Debian or Ubuntu example:
+#. To check that you have set these variables correctly in your current
+   environment, follow these example shell sessions (the
+   :envvar:`CROSS_COMPILE` value may be different on your system):
```
Even more verbose the 3rd time.

marc-hb commented on doc/getting_started/toolchain_host.rst:
```diff
@@ -0,0 +1,12 @@
+.. _host_toolchains:
```
That's a very short page. Is it included or referenced anywhere?

carlescufi replied:
This is already the case:
https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/16064/docs/getting_started/toolchain_host.html

mbolivar replied:
Neat! I wish the "details" link in the CI checks linked there instead of 404ing ;)

marc-hb commented on doc/getting_started/index.rst:
```diff
@@ -128,77 +95,36 @@ Next, install additional Python packages required by Zephyr in a shell or
-   On Linux, you can skip this step if you installed the :ref:`Zephyr SDK
-   <zephyr_sdk>`, which includes toolchains for all supported Zephyr
-   architectures.
+You *configure* the toolchain by setting environment variables. You need to set
```
I think you mean "select" the toolchain instead of "configure" the toolchain. The latter sounds like changing the way the toolchain behaves.

mbolivar replied:
I don't think so. ZEPHYR_TOOLCHAIN_VARIANT could maybe be described as "selecting" a toolchain, but the other variables which set its location are surely configuration options. For example, I keep several old Zephyr SDKs installed on my system, and configure the one which is currently in use based on how old the Zephyr tree I'm building against is.

marc-hb replied:
I'm not a native speaker (@dbkinder help?) but I understand that choosing whether to use SDK 0.9.5 or SDK 0.10.0 is configuring the Zephyr build, not "configuring the toolchain". Which one by the way? Both?
Configuring one toolchain would be changing its behaviour in some way and while possible I haven't found any example of that in the current documentation; it belongs to toolchain-specific documentation.

mbolivar replied:
I still disagree FWIW.

marc-hb commented on doc/getting_started/index.rst:
```diff
@@ -128,77 +95,36 @@ Next, install additional Python packages required by Zephyr in a shell or
 
-.. note::
+Zephyr binaries are compiled and linked by a *toolchain*\
+[#tools_native_posix]_. You now need to *install* and *configure* a toolchain.
```
From the perspective of "Getting Started as fast as possible", the native posix and host toolchains options are awesome because they _skip_ completely a big and time-consuming step: no need to install any SDK. That's why these options were highlighted this early. I could be wrong but I'm afraid you're making these options much easier to miss.

carlescufi replied:
The problem is that those only apply to Linux users, and also many Zephyr users will want to get up and running on real hardware as soon as possible, this being an embedded RTOS. They probably want to build, flash and run a simple app on a board they have.

mbolivar replied:
> That's why these options were highlighted this early.

As Carles says, those options only apply to Linux users. The Linux reader who is going through this document in linear order has already installed the Zephyr SDK, since that's in installation_linux.rst, or chosen to go against its recommendations. So: this section is skippable by those who follow recommendations, and those who don't have already decided they know better than us and can be trusted to read the details.

> I could be wrong but I'm afraid you're making these options much easier to miss.

I'm not so sure.

I've watched over the shoulder as experienced Linux users followed this guide; several of them skipped reading any and all text and just started blindly copy/pasting the first terminal commands they could find. So less may be more here.

Further, I think that having a choice which clearly says "Host Toolchains" is a pretty clear signal to this sort of user that they have choices for skipping that option.

marc-hb replied:
> The Linux reader who is going through this document in linear order has already installed the Zephyr SDK, since that's in installation_linux.rst, or chosen to go against its recommendations

You're right, I missed that. Much easier to spot now that I know about https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/16064/docs/getting_started/index.html !

carlescufi reviewed (approved):
LGTM now, thank you!

mbolivar-nordic committed:
doc: clear up qemu + zephyr sdk interaction

Make it clearer that the note about setting QEMU_BIN_PATH only applies
if you've got the Zephyr SDK installed.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar commented:
v5 is up; I've responded to all comments.

dbkinder reviewed (changes requested):
Nice update all!

ulfalizer commented on doc/getting_started/toolchain_custom_cmake.rst:
```diff
@@ -3,25 +3,29 @@
+- :file:`cmake/toolchain/generic.cmake`: configures the toolchain for "generic"
+  use (mostly to run the C preprocessor on the generated :ref:`device-tree`
+  file).
+- :file:`cmake/toolchain/target.cmake`: configures the toolchain for use
```
Extra "use".

mbolivar replied:
Nope. Maybe you're reading "use" as a verb here, not a noun?

dbkinder replied:
Hmmm... what does "configures the toolchain for use building Zephyr" mean?  I think the "use" should be deleted too.

mbolivar replied:
It means configuring the toolchain so that it may be used to build zephyr applications...

marc-hb replied:
*This* "configure" I understand :-)

dbkinder commented:
@mbolivar can we finish the edits to get this merged...

mbolivar commented on doc/getting_started/index.rst:
```diff
@@ -29,85 +29,53 @@ Follow one of the following guides for your host operating system.
-   It's possible to use Zephyr without installing west, but you have
-   to **really** know :ref:`what you are doing <no-west>`.
+Zephyr's multi-purpose :ref:`west <west>` tool lets you easily get the Zephyr
+project repositories. (It's possible to use Zephyr without installing west, but
```
Hmm... that's longer, not shorter (the main goal of this PR is to make it shorter and easier to read), and the sentence I think you find unkind ("It's possible ... really know what you're doing") is already part of this document.

I personally don't take that sentence as unkind (though I wrote it originally, so grain of salt there). I also disagree with the suggestion; experts are experts and I certainly wouldn't make that recommendation to one.

To try to preserve the meaning without making it longer, I'll switch to:

```
(It's possible to use Zephyr without installing west, but
:ref:`that's documented elsewhere <no-west>`.)
```

mbolivar replied:
How about "While it's easiest to develop with Zephyr using west, we also show the ..."?

West makes a lot of things much more convenient, but there is a small, vocal, and considered minority with valid use cases to not use it. So for me personally authoring this PR and showing up in the `git blame` lines, "recommend" is not the word I would use. Feel free to submit your own changes making recommendations, of course :).

mbolivar replied:
> I think the point was to say using west is the recommended approach.

Changed to this:

```
While it's easiest to develop with Zephyr using west,
we also have :ref:`documentation for doing without it <no-west>`.
``

marc-hb replied:
Either "it's easiest" or "we recommend" is fine, they mean pretty much the same thing unless you're the person who has been receiving a lot "constructive" feedback from a vocal minority of people trying to avoid west.

"Make simple things [west] simple, and complex things [no west] possible"

Don't let people complaining that "the complex is not easy" affect you. Just keep it "possible" that's all :-)

PS: this thread is not available in the "Files Changed" tab any more. I think it's (again) because of force-pushes.

mbolivar replied:
As Rich Hickey eloquently argued in Simple Made Easy (https://www.youtube.com/watch?v=34_L7t7fD_U), easiest isn't always best :smile:

mbolivar commented:
> @mbolivar can we finish the edits to get this merged...

Sorry for delay, now done

dbkinder commented on doc/getting_started/index.rst:
```diff
@@ -29,85 +29,53 @@ Follow one of the following guides for your host operating system.
+  require using this option.
 
-   You should see ``West bootstrapper version: v0.5.0`` (or higher).
+Clone the Zephyr Repositories
```
Need to add the ``.. _clone-zephyr:`` back before this this title (from line 57 in the original file).  This label is referenced from the guides/west/install.rst file (that's the cause of the doc build error you're getting)

mbolivar replied:
Whoops, sorry, thanks!

dbkinder commented on doc/getting_started/index.rst:
```diff
@@ -128,85 +95,36 @@ Next, install additional Python packages required by Zephyr in a shell or
-   environment variable set to the Zephyr SDK installation directory, while
-   setting :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` appropriately for a non-SDK
-   toolchain.
+You *configure* the toolchain by setting :ref:`environment variables
```
```suggestion
You *configure* the toolchain to use by setting :ref:`environment variables
```

marc-hb commented on doc/getting_started/toolchain_custom_cmake.rst:
```diff
@@ -3,25 +3,29 @@
-You can also set them as CMake variables when generating a build
-system for a Zephyr application, like so:
+You can also set ``ZEPHYR_TOOLCHAIN_VARIANT`` and ``TOOLCHAIN_ROOT`` as CMake
+variables when generating a build system for a Zephyr application, like so:
```
A bit "late to the party"; just realized it would be great to add a small `cmake -C project1_params.cmake` example here. Because `-C` is terribly misnamed "preload cache" in cmake documentation I kept missing it. Eventually I found it's perfect for doing exactly this while scaling so much better: long lists, version-controlled configurations, etc.

mbolivar replied:
I've added a note in a follow-on commit touching just this file.

marc-hb reviewed (commented):
I re-iterate my suggestion to please "divide and conquer" and split such large PRs in several, smaller PRs in the future. For instance I'm not going to look at the latest revision of this PR because it's just too big and there's no way to compare it with the previous force push.

dbkinder reviewed (approved):
I still don't like "use building", but let's get this merged and move on  :)

dbkinder commented on doc/getting_started/index.rst:
```diff
@@ -29,85 +29,55 @@ Follow one of the following guides for your host operating system.
+  require using this option.
 
-      west --version
+.. _clone-zephyr::
```
Whoops, only one colon at the end of the line (that's generating the malformed label error)

dbkinder reviewed (changes requested):
doc build error fix

mbolivar-nordic committed:
getting_started: changes for custom toolchains

This option uses a lot of typing, so point out there's a way to save
keystrokes using -C.

Tweak the language around the target toolchain.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar commented:
> I still don't like "use building", but let's get this merged and move on :)

It's not "use building".

It's "configuring for use", similarly to how "conditions for use" appears in the Apache license text: https://github.com/zephyrproject-rtos/zephyr/blob/master/LICENSE#L5

Since this is bothering multiple people, I've reworded it.

mbolivar commented:
>  re-iterate my suggestion to please "divide and conquer" and split such large PRs in several, smaller PRs in the future.

I try to make commits self-contained. If you have specific suggestions for how I could have structured these commits differently to make them easier to review, I would appreciate hearing them.

marc-hb commented:
> I try to make commits self-contained. If you have specific suggestions for how I could have structured these commits differently to make them easier to review, I would appreciate hearing them.

I'm referring to PRs, not commits. This was 4 independent and/or large commits -> 4 different Pull Requests. Easy!

4 smaller PRs instead of a 1 large PR => smaller reviews + fewer force-pushes per PR => **faster** reviews => merged _sooner_!

Yes, I know github doesn't let you easily chain PRs depending on each other https://github.com/zephyrproject-rtos/zephyr/pull/14444#issuecomment-472236477
There are a few workarounds, all worth it: - All chained PRs except the first open one are [DNM] (some examples in #16031?); - only one 1 PR at a time and the others are just branches, published or not; - others?

dbkinder reviewed (approved):
Works for me, thanks!
